ggml-metal.metal

```diff
@@ -8,14 +8,14 @@ using namespace metal;
 #define QR4_0 2
 typedef struct {
     half    d;             // delta
-    uint8_t qs[QK4_0 / 2]; // nibbles / quants
+    uint16_t qs[QK4_0 / 4]; // nibbles / quants
```

Comment:
I personally do not like having the Metal `block_q4_0/1` definition differ from the definition in `ggml.c`. One can get the same effect by just casting `qs` to `uint16_t *` where needed.

Reply:
I agree that it's better to keep a uniform design across the library. However, changing `uint16_t` back to `uint8_t` and casting `qs`  to `uint16_t *` almost kills all the performance gain. 

Disassembler shows that the complier is just not clever enough to load 16 `uint8_t`  in a reasonable way. It will first load 2 Bytes, then 8 Bytes, then all the left Bytes, taking 3~4 load instructions. It does load 8 `uint16_t` in 2 instructions, with each instructions loading 8 Bytes. 

The saved `mov` instructions by using `uint16_t` should benefit more on future matrix-matrix multiplication kernel, for that is more compute bound.

Reply:
> However, changing uint16_t back to uint8_t and casting qs to uint16_t * almost kills all the performance gain.
>
> Disassembler shows that the complier is just not clever enough to load 16 uint8_t in a reasonable way. It will first load 2 Bytes, then 8 Bytes, then all the left Bytes, taking 3~4 load instructions. It does load 8 uint16_t in 2 instructions, with each instructions loading 8 Bytes.

Surprising. I also would have preferred to keep the structs the same, but if there is no way to workaround this I guess we can make the change.

Can you give a very short tutorial for beginners how to generate and look at the disassembler output?

Reply:
Hi @ggerganov, A quick guide is post at #2279.